hack/update-codegen.sh

```diff
@@ -43,20 +43,38 @@ function generateApplyConfiguration(){
       done
     done
 
-    echo "Generating applyconfigurations"
+   if [ "$OUTPUT_PKG" == "github.com/openshift/client-go/machineconfiguration" ] 
+   then
+    echo "Generating applyconfigurations specifically for MCO"
     applyconfigurationgen_external_apis_csv="$(codegen::join , "${FQ_APIS[@]}")"
     applyconfigurations_package="${OUTPUT_PKG}/${CLIENTSET_PKG_NAME:-applyconfigurations}"
     ${APPLYCONFIGURATION_GEN}  \
       --output-package "${applyconfigurations_package}" \
       --input-dirs "${applyconfigurationgen_external_apis_csv}" \
+      --external-applyconfigurations k8s.io/api/core/v1.ObjectReference:k8s.io/client-go/applyconfigurations/core/v1 \
```

Comment:
So the MCO embeds a `corev1.ObjectReference` [here](https://github.com/openshift/client-go/blob/bf3c41078b4c840c32692dce23680d123f7e8ed8/vendor/github.com/openshift/api/machineconfiguration/v1/types.go#L425), which makes applyconfiguration-gen generate code that doesn't compile. 

It looks like by default, `applyconfiguration-gen` includes the following as external by default (which I believe is why this isn't an issue with the `metav1.TypeMeta/metav1.ObjectMeta` that seem to work: 
```
k8s.io/apimachinery/pkg/apis/meta/v1.TypeMeta:k8s.io/client-go/applyconfigurations/meta/v1,

k8s.io/apimachinery/pkg/apis/meta/v1.ObjectMeta:k8s.io/client-go/applyconfigurations/meta/v1,

k8s.io/apimachinery/pkg/apis/meta/v1.OwnerReference:k8s.io/client-go/applyconfigurations/meta/v1
```

If I add `corev1.ObjectReference` as external, and point it to its appyconfiguration: 
```
k8s.io/api/core/v1.ObjectReference:k8s.io/client-go/applyconfigurations/core/v1 
```
Everything generates right for the MCO, but it changes a bunch of the other packages that were originally generated without the presence of `v1.ObjectReferenceApplyConfiguration `, and that seems...not good...because they've already been working well for quite some time.

Reply:
It's curious that the MCO blows up but others haven't, is there any difference in how the object reference is being used in the other APIs? Pointer vs reference per chance?

Reply:
There are examples in the API of both pointer/non-pointer usages, but ours is the *only* one that is embedded:
```
apps/v1/types.go:       To corev1.ObjectReference `json:"to" protobuf:"bytes,2,opt,name=to"`
apps/v1/types.go:       From corev1.ObjectReference `json:"from" protobuf:"bytes,3,opt,name=from"`
apps/v1/types.go:       From corev1.ObjectReference `json:"from" protobuf:"bytes,1,opt,name=from"`
apps/v1/types.go:       From corev1.ObjectReference `json:"from" protobuf:"bytes,1,opt,name=from"`
authorization/v1/types.go:      Subjects []corev1.ObjectReference `json:"subjects" protobuf:"bytes,4,rep,name=subjects"`
authorization/v1/types.go:      RoleRef corev1.ObjectReference `json:"roleRef" protobuf:"bytes,5,opt,name=roleRef"`
authorization/v1/types.go:      Subjects []corev1.ObjectReference `json:"subjects" protobuf:"bytes,4,rep,name=subjects"`
authorization/v1/types.go:      RoleRef corev1.ObjectReference `json:"roleRef" protobuf:"bytes,5,opt,name=roleRef"`
build/v1/types.go:      FromRef *corev1.ObjectReference `json:"fromRef,omitempty" protobuf:"bytes,2,opt,name=fromRef"`
build/v1/types.go:      Config *corev1.ObjectReference `json:"config,omitempty" protobuf:"bytes,9,opt,name=config"`
build/v1/types.go:      From corev1.ObjectReference `json:"from" protobuf:"bytes,1,opt,name=from"`
build/v1/types.go:      From corev1.ObjectReference `json:"from" protobuf:"bytes,1,opt,name=from"`
build/v1/types.go:      From *corev1.ObjectReference `json:"from,omitempty" protobuf:"bytes,1,opt,name=from"`
build/v1/types.go:      From corev1.ObjectReference `json:"from" protobuf:"bytes,1,opt,name=from"`
build/v1/types.go:      To *corev1.ObjectReference `json:"to,omitempty" protobuf:"bytes,1,opt,name=to"`
build/v1/types.go:      From *corev1.ObjectReference `json:"from,omitempty" protobuf:"bytes,2,opt,name=from"`
build/v1/types.go:      TriggeredByImage *corev1.ObjectReference `json:"triggeredByImage,omitempty" protobuf:"bytes,3,opt,name=triggeredByImage"`
build/v1/types.go:      From *corev1.ObjectReference `json:"from,omitempty" protobuf:"bytes,4,opt,name=from"`
image/v1/types.go:      From *corev1.ObjectReference `json:"from,omitempty" protobuf:"bytes,3,opt,name=from"`
image/v1/types.go:      From corev1.ObjectReference `json:"from" protobuf:"bytes,1,opt,name=from"`
image/v1/types.go:      From corev1.ObjectReference `json:"from" protobuf:"bytes,1,opt,name=from"`
machine/v1beta1/types_machinehealthcheck.go:    RemediationTemplate *corev1.ObjectReference `json:"remediationTemplate,omitempty"`
machine/v1beta1/types_machine.go:       NodeRef *corev1.ObjectReference `json:"nodeRef,omitempty"`
security/v1/types.go:   AllowedBy *corev1.ObjectReference `json:"allowedBy,omitempty" protobuf:"bytes,1,opt,name=allowedBy"`
template/v1/types.go:   Ref corev1.ObjectReference `json:"ref,omitempty" protobuf:"bytes,1,opt,name=ref"`
template/v1/types.go:   TemplateInstance corev1.ObjectReference `json:"templateInstance" protobuf:"bytes,1,opt,name=templateInstance"`
template/v1/types.go:   Secret corev1.ObjectReference `json:"secret" protobuf:"bytes,2,opt,name=secret"`
user/v1/types.go:       User corev1.ObjectReference `json:"user" protobuf:"bytes,4,opt,name=user"`
user/v1/types.go:       Identity corev1.ObjectReference `json:"identity,omitempty" protobuf:"bytes,2,opt,name=identity"`
user/v1/types.go:       User corev1.ObjectReference `json:"user,omitempty" protobuf:"bytes,3,opt,name=user"`
```
vs
```
machineconfiguration/v1/types.go:       PullSecret *corev1.ObjectReference `json:"pullSecret,omitempty" protobuf:"bytes,9,opt,name=pullSecret"`
---> machineconfiguration/v1/types.go:       corev1.ObjectReference `json:",inline"  protobuf:"bytes,1,rep,name=objectReference"`
machineconfiguration/v1/types.go:       Source []corev1.ObjectReference `json:"source,omitempty" protobuf:"bytes,2,rep,name=source"`
```

Reply:
Ahh, right, in general for APIs we try to avoid embedding structs like that, but, it's done now.

In theory you could change the go type without causing serialization errors, copying the fields from the object reference into the parent struct, that would mean you don't have this hack any more.

Also, we should remove the protobuf tags, they aren't used for CRDs

Reply:
> In theory you could change the go type without causing serialization errors, copying the fields from the object reference into the parent struct, that would mean you don't have this hack any more.

Something like this? https://gist.github.com/jkyros/db4d27fb28b38fa2ea170164ebfabae9#file-hypothetical_atrocity-go That feels like it should be a crime :stuck_out_tongue:  I think it would work, but there is stuff in the MCO that cares that it's an ObjectReference (mostly tests)  and we'd need to defuse all of that. 

Would you entertain letting us keep this MCO-only applyconfig hack for now (or something similar -- I don't want to make your update script ugly)  until we can do some rearrangement? The applyconfigurations generated by this do seem to be proper and work, e.g. https://gist.github.com/jkyros/db4d27fb28b38fa2ea170164ebfabae9#file-main-go

Reply:
Yeah the gist looks like what I meant. Since the serialization is the same, it's fine for a CRD to make a change like that to the go type, I got second opinions on that and a few of us are in agreement 😅 

I'm happy to let you carry on like this, but, can we stick a pin in this to fix it before the end of 4.15? I'd like a TODO to remove this hack